### PR TITLE
implementing advancedcronjob controller

### PR DESCRIPTION
**Pull request by rishi-anand**
### Ⅰ. Describe what this PR does
This PR implements controller for AdvancedCronJob. 
Proposal: https://github.com/openkruise/kruise/blob/master/docs/proposals/20201024-advancedcronjob.md

### Ⅱ. Does this pull request fix one issue?

#215 

### Ⅲ. List the added test cases (unit test/integration test) if any, please explain if no tests are needed.
Will be added later

### Ⅳ. Describe how to verify it
WIll add this information after implementing webhook

### Ⅴ. Special notes for reviews
@Fei-Guo, Implemented controller in this PR. In next PR, I will add webhook.

### CODE REVIEW

File: pkg/controller/advancedcronjob/advancedcronjob_controller.go [hunk 1]
```diff
@@ -135,18 +131,32 @@ type ReconcileAdvancedCronJob struct {
 
 func (r *ReconcileAdvancedCronJob) Reconcile(req ctrl.Request) (ctrl.Result, error) {
 	ctx := context.Background()
+	klog.Infof("Running BroadcastCronJob job %s", req.Name)
```
Comment (Fei-Guo) [inline pkg/controller/advancedcronjob/advancedcronjob_controller.go:134]: BroadcastCronJob -> AdvancedCronJob

File: pkg/util/fieldindex/register.go [hunk 2]
```diff
@@ -71,6 +64,79 @@ func RegisterFieldIndexes(c cache.Cache) error {
...
+		}
+		// broadcastjob owner
+		if err = indexBroadcastCronJob(c); err != nil {
+			return
+		}
 	})
 	return err
 }
+
+func indexPodName(c cache.Cache) error {
```
Comment (Fei-Guo) [inline pkg/util/fieldindex/register.go:83]: Why do we need to create a field index for pod name?

File: pkg/util/fieldindex/register.go [hunk 3]
```diff
@@ -71,6 +64,79 @@ func RegisterFieldIndexes(c cache.Cache) error {
...
+		// ...and if so, return it
+		return []string{owner.Name}
+	}); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func indexBroadcastCronJob(c cache.Cache) error {
```
Comment (PualrDwade) [inline pkg/util/fieldindex/register.go:122]: how about change the style just like:
```go
return  c.IndexField(&appsv1alpha1.BroadcastJob{}, IndexNameForController, func(rawObj runtime.Object) []string {
		// grab the job object, extract the owner...
		job := rawObj.(*appsv1alpha1.BroadcastJob)
		owner := metav1.GetControllerOf(job)
		if owner == nil {
			return nil
		}

		// ...make sure it's a AdvancedCronJob...
		if owner.APIVersion != apiGVStr || owner.Kind != appsv1alpha1.AdvancedCronJobKind {
			return nil
		}

		// ...and if so, return it
		return []string{owner.Name}
	})
```

File: pkg/util/fieldindex/register.go (hunk 2 above)
Comment (rishi-anand) [reply pkg/util/fieldindex/register.go:83]: That was the existing code. Just formatted it in different func

File: pkg/util/fieldindex/register.go (hunk 3 above)
Comment (rishi-anand) [reply pkg/util/fieldindex/register.go:115]: Done in commit https://github.com/openkruise/kruise/pull/461/commits/3d513814466ce219a7543280c4efc3a697a9fe78

File: pkg/controller/advancedcronjob/advancedcronjob_controller.go (hunk 1 above)
Comment (rishi-anand) [reply pkg/controller/advancedcronjob/advancedcronjob_controller.go]: Done in commit https://github.com/openkruise/kruise/pull/461/commits/3d513814466ce219a7543280c4efc3a697a9fe78

File: pkg/controller/advancedcronjob/advancedcronjob_job_controller.go [hunk 4]
```diff
@@ -0,0 +1,378 @@
+package advancedcronjob
```
[the remark this answers is not in the document]
Comment (rishi-anand) [reply pkg/controller/advancedcronjob/advancedcronjob_job_controller.go:15]: Done in commit https://github.com/openkruise/kruise/pull/461/commits/3d513814466ce219a7543280c4efc3a697a9fe78

File: pkg/controller/advancedcronjob/advancedcronjob_broadcastjob_controller.go [hunk 5]
```diff
@@ -0,0 +1,376 @@
+package advancedcronjob
```
[the remark this answers is not in the document]
Comment (rishi-anand) [reply pkg/controller/advancedcronjob/advancedcronjob_broadcastjob_controller.go:15]: Done in commit https://github.com/openkruise/kruise/pull/461/commits/3d513814466ce219a7543280c4efc3a697a9fe78

File: pkg/util/fieldindex/register.go (hunk 2 above)
Comment (Fei-Guo) [reply pkg/util/fieldindex/register.go:83]: Please rename to indexPodNodeName, which is different from creating an index for Pod name.

**Comment by Fei-Guo** [comment]
Please address above comment. Otherwise, LGTM.

File: pkg/util/fieldindex/register.go (hunk 2 above)
Comment (rishi-anand) [reply pkg/util/fieldindex/register.go]: Addressed in https://github.com/openkruise/kruise/pull/461/commits/5b783be22bff9600f7f0cb820dc55ef2b64998fd

**Comment by rishi-anand** [comment]
> Please address above unresolved comment. Otherwise, LGTM.

Addressed in https://github.com/openkruise/kruise/pull/461/commits/5b783be22bff9600f7f0cb820dc55ef2b64998fd

**Review by Fei-Guo — APPROVED** [review]

**Comment by FillZpp** [comment]
@rishi-anand Could you please add some unit tests for advancedcronjob controller?

**Comment by rishi-anand** [comment]
@FillZpp, Sure will add test cases..

**Pull request merged**